cwltool/workflow.py

```diff
@@ -93,31 +94,73 @@ def match_types(sinktype, src, iid, inputobj, linkMerge, valueFrom):
     return False
 
 
-def can_assign_src_to_sink(src, sink):  # type: (Any, Any) -> bool
+def check_types(srctype, sinktype, linkMerge, valueFrom):
+    # type: (Union[List[Text],Text], Union[List[Text],Text], Text, Text) -> Text
+    """Check if the source and sink types are "pass", "warning", or "exception".
+    """
+
+    if valueFrom:
+        return "pass"
+    elif not linkMerge:
+        if can_assign_src_to_sink(srctype, sinktype, strict=True):
+            return "pass"
+        elif can_assign_src_to_sink(srctype, sinktype, strict=False):
+            return "warning"
+        else:
+            return "exception"
+    else:
+        if not isinstance(sinktype, dict):
+            return "exception"
+        elif linkMerge == "merge_nested":
+            return check_types(srctype, sinktype["items"], None, None)
+        elif linkMerge == "merge_flattened":
+            if not isinstance(srctype, dict):
+                return check_types(srctype, sinktype["items"], None, None)
+            else:
+                return check_types(srctype, sinktype, None, None)
+        else:
+            raise WorkflowException(u"Unrecognized linkMerge enum '%s'" % linkMerge)
+
+
+def can_assign_src_to_sink(src, sink, strict=False):  # type: (Any, Any, bool) -> bool
     """Check for identical type specifications, ignoring extra keys like inputBinding.
+
+    src: admissible source types
+    sink: admissible sink types
+
+    In non-strict comparison, at least one source type must match one sink type.
+    In strict comparison, all source types must match one sink type.
     """
+
     if sink == "Any":
         return True
     if isinstance(src, dict) and isinstance(sink, dict):
         if src["type"] == "array" and sink["type"] == "array":
-            return can_assign_src_to_sink(src["items"], sink["items"])
+            return can_assign_src_to_sink(src["items"], sink["items"], strict)
         elif src["type"] == "record" and sink["type"] == "record":
-            return _compare_records(src, sink)
+            return _compare_records(src, sink, strict)
```

Comment:
Because you removed "return False" at the bottom of the function, you need to add it here, otherwise if src["type"] != sink["type"] it'll fall through and return None instead of False.